### DISCUSSION TITLE
[Proposal] Alternative type alias: typedef

### DISCUSSION THREAD

**Discussion by kinderhead** [dotnet/csharplang#6311] [Language Ideas]
# FEATURE_NAME

* [x] Proposed
* [ ] Prototype: Not Started
* [ ] Implementation: Not Started
* [ ] Specification: Not Started

## Summary
[summary]: #summary

This has probably been suggested in this way somewhere, but here we are. I propose a C++ `typedef` like feature for type aliases.

## Motivation
[motivation]: #motivation

Type aliases are useful when needing to use the same generic class multiple times. Using `using` type aliases can be a hassle if the same alias is needed across a larger project. A solution is `typedef`. The `typedef` statement can be used as type alias for all who are `using` the namespace in which the typedef is in.

## Detailed design
[design]: #detailed-design

Example:
``` C#
// File1.cs
namespace Showcase {
    public struct Vector2<T> {
        public T X;
        public T Y;

        public Vector2(T x, T y) {
            X = x;
            Y = y;
        }

        // ...
    }

    typedef Vector2i Vector2<int>;
}

// File2.cs
using Showcase;

// ...
public void Function(Vector2i vec) {
    DoSomething(vec.X + 50);
}
```

Now one may be saying, "you could just do `public struct Vector2i : Vector2<int>`." Structs cannot have inheritance, and interfaces can be limiting sometimes or not desireable in certain scenarios.

Other examples:
``` C#
typedef ShortName<T> AVeryLongNameThatIsProbablyUnnecessary<T>;
typedef Hello<T> World<T, int, string, ShortName<T>>;
```

`typedef` inherits the access modifier of the aliased type, so a `typedef` of `private class Private` will be private. It is only a compile time construct. It is substituted before code generation.

Another idea is that delegates can be generated from using `typedef` with a function.
``` C#
public static class Baz {
    public static void Foo(string message) {
        Console.Writeline(message);
    }
}

typedef Bar Baz.Foo;
```

This would replace the `typedef` statement with `public delegate void Bar(string message);`

## Drawbacks
[drawbacks]: #drawbacks

## Alternatives
[alternatives]: #alternatives

You could just use `global using`.

## Unresolved questions
[unresolved]: #unresolved-questions

**Comment by TahirAhmadov** [comment]
Is this what you have in mind?
https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-10.0/globalusingdirective

**Reply by FaustVX** [reply]
Combined with #1239 will be great

**Reply by kinderhead** [reply]
I actually didn't know about that. What I have in mind is a bit different though

**Comment by huoyaoyuan** [comment]
Duplicate of #1300
